text/0000-crt-static.md

```diff
@@ -0,0 +1,371 @@
+export CARGO_CFG_DEBUG_ASSERTIONS
+```
+
+As mentioned in the previous section, the linkage of the C standard library will
+be specified as a target feature, which is lowered to a `cfg` value, thus giving
+build scripts the ability to modify compilation options based on C standard
+library linkage. One important complication here is that `cfg` values in Rust
+may be defined multiple times, and this is the case with target features. When a
+`cfg` value is defined multiple times, Cargo will create a single environment
+variable with a comma-separated list of values.
+
+So for a target with the following features enabled
+
+```
+target_feature="sse"
+target_feature="crt-static"
+```
+
+Cargo would convert it to the following environment variable:
+
+```
+export CARGO_CFG_TARGET_FEATURE=sse,crt-static
+```
+
+Through this method build scripts will be able to learn how the C standard
+library is being linked.  This is crucially important for the MSVC target where
+code needs to be compiled differently depending on how the C library is linked.
+
+This feature ends up having the added benefit of informing build scripts about
+selected CPU features as well. For example once the `target_feature` `#[cfg]`
+is stabilized build scripts will know whether SSE/AVX/etc are enabled features
+for the C code they might be compiling.
+
+After this change, the gcc-rs crate will be modified to check for the
+`CARGO_CFG_TARGET_FEATURE` directive, and parse it into a list of enabled
+features. If the `crt-static` feature is not enabled it will compile C code on
+the MSVC target with `/MD`, indicating dynamic linkage. Otherwise if the value
+is `static` it will compile code with `/MT`, indicating static linkage. Because
+today the MSVC targets use dynamic linkage and gcc-rs compiles C code with `/MD`,
+gcc-rs will remain forward and backwards compatible with existing and future
+Rust MSVC toolchains until such time as the the decision is made to change the
+MSVC toolchain to `+crt-static` by default.
+
+### Lazy link attributes
+
+The final feature that will be added to the compiler is the ability to "lazily"
+interpret the linkage requirements of a native library depending on values of
+`cfg` at compile time of downstream crates, not of the crate with the `#[link]`
+directives. This feature is never intended to be stabilized, and is instead
+targeted at being an unstable implementation detail of the `libc` crate linked
+to `std` (but _not_ the stable `libc` crate deployed to crates.io).
+
+Specifically, the `#[link]` attribute will be extended with a new argument
+that it accepts, `cfg(..)`, such as:
+
+```rust
+#[link(name = "foo", cfg(bar))]
+```
+
+This `cfg` indicates to the compiler that the `#[link]` annotation only applies
+if the `bar` directive is matched. This interpretation is done not during
+compilation of the crate in which the `#[link]` directive appears, but during
+compilation of the crate in which linking is finally performed. The compiler
```

Comment:
@japaric yes the solution for musl here will be a tricky one, but it's intended to just be an implementation detail. I'm thinking that we'll detect that the statically included library does not need to be linked (due to how `#[link(..., cfg(...))]` was evaluated) which will cause the compiler to create a new temporary rlib. This temporary rlib (like those we create for LTO) won't have all the original contents, just the Rust object file.

It won't be perfect, but it'll solve the precisely one use case we have for this attribute today (and is another reason why the attribute will remain unstable).